src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+                    Trim(tfl.Location);
+
+                    if (tfl.Index == -1)
+                    {
+                        if (!_uniqueThreadFlows.TryGetValue(tfl, out int tflIndex))
+                        {
+                            tflIndex = _run.ThreadFlowLocations.Count;
+                            _run.ThreadFlowLocations.Add(tfl);
+                            _uniqueThreadFlows[tfl] = tflIndex;
+                        }
+
+                        tFlow.Locations[i] = new ThreadFlowLocation() { Index = tflIndex };
+                    }
+                }
+            }
+        }
+
+        public IList<Location> Trim(IList<Location> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<Location> uniqueLocations = new HashSet<Location>(Location.ValueComparer);
+            List<Location> newLocations = new List<Location>();
+
+            this.TotalLocations += locations.Count;
+            for (int i = 0; i < locations.Count; ++i)
+            {
+                Location loc = locations[i];
+
+                // Trim Locations themselves
+                Trim(loc);
+
+                // Keep only one copy of each unique location
+                if (uniqueLocations.Add(loc))
+                {
+                    this.UniqueLocations++;
+                    newLocations.Add(loc);
+                }
+            }
+
+            return newLocations;
+        }
+
+        public void Trim(Location loc)
+        {
+            if (loc != null)
+            {
+                // Remove Ids
+                loc.Id = -1;
+
+                Trim(loc.PhysicalLocation);
+                Trim(loc.LogicalLocation);
+                Trim(loc.LogicalLocations);
+            }
+        }
+
+        public IList<LogicalLocation> Trim(IList<LogicalLocation> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<LogicalLocation> uniqueLocations = new HashSet<LogicalLocation>(LogicalLocation.ValueComparer);
+            List<LogicalLocation> newLocations = new List<LogicalLocation>();
+
+            this.TotalLocations += locations.Count;
+            for (int i = 0; i < locations.Count; ++i)
+            {
+                LogicalLocation loc = locations[i];
+
+                // Trim Locations themselves
+                Trim(loc);
+
+                // Keep only one copy of each unique location
+                if (uniqueLocations.Add(loc))
+                {
+                    this.UniqueLocations++;
+                    newLocations.Add(loc);
+                }
+            }
+
+            return newLocations;
+        }
+
+        public void Trim(LogicalLocation loc)
+        {
+            if (loc != null)
+            {
+                // Leave only index if index is present
+                if (loc.Index != -1)
+                {
+                    loc.DecoratedName = null;
```

Comment:
>loc.DecoratedName [](start = 20, length = 17)

what's our guarantee here and elsewhere that the object pointed to in the run level array contains all this information? ie, if we zap FullyQualifiedName from the result associated logical location, we should be sure that it is expressed in the run.logicalLocations instance pointed to by Index. all this points to an unpleasant inconsistency that could exist in SARIF (for which we don't have a validator), SARIF results could themselves point to a common array resident thing while being inconsistent in its local recapitulation of properties.

Reply:
Yes, the spec addresses that issue. For example, in §3.33.3 (logicalLocation.index property):

    If `index` is present, `thisObject` **SHALL** take all properties present on the cached object. If `thisObject` contains any properties other than `index`, they **SHALL** equal the corresponding properties of the cached object.

I filed [https://github.com/microsoft/sarif-sdk/issues/1674](https://github.com/microsoft/sarif-sdk/issues/1674), "New rule: Ensure that local objects are consistent with cached objects". 

---
In reply to: [324280903](https://github.com/microsoft/sarif-sdk/pull/1672#discussion_r324280903) [](ancestors = 324280903)

Reply:
That will be a big job, comparable in size to the implementation of `SARIF1017`, "Invalid index".

---
In reply to: [324342614](https://github.com/microsoft/sarif-sdk/pull/1672#discussion_r324342614) [](ancestors = 324342614,324280903)

Reply:
I really don't think it should be the SarifConsolidator's job to ensure every property from the ArtifactLocation is copied to the Run.Artifact when an index has been specified already by the producer. The consolidator should omit other properties when an index has been specified, because all of them are redundant, but I think that's a reasonable boundary for what it does.